### ISSUE TITLE
Grid interpolation explanations?

### ISSUE THREAD

**Issue by angus-g** [OceanParcels/parcels#769]
It seems a bit unclear how data should be structured for different grid interpolation schemes (for B-grid and C-grid data, in particular). For example, I gather that B-grid data expects U and V to be vertically-constant within a cell, and W to be laterally-constant. However, we usually see W as vertically staggered. Indeed, parcels specifically checks that the U/V and W vertical grids are the same.

How should we unify these two interpretations?

**Comment by erikvansebille** [comment]
This is something that @delandmeterp discusses in our paper at https://www.geosci-model-dev.net/12/3571/2019/gmd-12-3571-2019.html (section 2.1.3). Does that help?

**Comment by angus-g** [comment]
Sorry to reopen this one, but I feel like the description (and my understanding) still doesn't match up with the code:

> For a three-dimensional cell, the tracer node is still at the centre of the cell and the field is constant; the four horizontal velocity  nodes  are  located  at  the  middle  of  the  four  vertical edges; the two vertical velocity nodes are located at the centre of the two horizontal faces

I have U, V data with dimensions `(xu_ocean, yu_ocean, st_ocean)`, placing it at the middle of the vertical edges of cells as expected. The W data is `(xt_ocean, yt_ocean, sw_ocean)`, which puts it at the middle of the top/bottom faces of cells. So far, I think this in accordance with a B-grid, and the description in the paper.

However, using `FieldSet.from_b_grid_dataset` does the following check, which suggests that all the data should be colocated for some reason:

```python
        if 'U' in dimensions and 'W' in dimensions and dimensions['U'] != dimensions['W']:
            raise RuntimeError("On a B-grid discretisation, U, V and W should have the same dimensions")
```

**Comment by erikvansebille** [comment]
Sorry for the confusion, @angus-g. This is really @delandmeterp's expertise, but I'll try to explain the way B-grid (and C-grid) interpolation works as best as I can

The point is that Parcels knows what a B-grid and a C-grid is, and has specific ways to interpolate these. See e.g. [this part of the FieldSet.from_c_grid_dataset() docstring](https://github.com/OceanParcels/parcels/blob/master/parcels/fieldset.py#L474-L490). Therefore, Parcels expects the locations of the f-nodes on the corners of the cell (see schematic from NEMO [here](https://www.nemo-ocean.eu/doc/img360.png)).

The same is true for B-grids (see [this part of the FieldSet.from_b_grid_dataset() docstring](https://github.com/OceanParcels/parcels/blob/master/parcels/fieldset.py#L617-L633). Parcels expects the locations of the corners of the grid cells. See for example [this line](https://github.com/OceanParcels/parcels/blob/master/parcels/examples/example_dask_chunk_OCMs.py#L70) where we load POP data:
```python
filenames = path.join(path.join(path.dirname(__file__), 'POPSouthernOcean_data'), 't.x1_SAMOC_flux.1690*.nc')
variables = {'U': 'UVEL', 'V': 'VVEL', 'W': 'WVEL'}
timestamps = np.expand_dims(np.array([np.datetime64('2000-%.2d-01' % m) for m in range(1, 7)]), axis=1)
dimensions = {'lon': 'ULON', 'lat': 'ULAT', 'depth': 'w_dep'}
```

So the check that you're referring to is intentional and important to prevent users from making mistakes in how they define the `dimensions` dictionary.

The alternative is to give each variable its own locations in the mesh, but then you'd have to use the normal `FieldSet.from_netcdf()` method, which assumes A-grid interpolation and would not be as accurate for B- or C-grid interpolation.

Makes sense? Very keen to hear any feedback on how we can clarify this better, since this is indeed quite important!

**Comment by angus-g** [comment]
I think the discrepancy in the explanation comes from showing that variables aren't located in the same place, but then expecting they all have the same coordinate data.

It seems like the missing part is the reasoning for wanting all the data to be _labelled_ at the same point, which in the case of a B-grid is the bottom(?) corner of a cell. This way when you get a parcel's sub-cell position, it's in the local (xsi, eta, zi) space for a given cell, instead of possibly ending up between cells.

Perhaps the confusion comes from the difference in thinking of the data in index-space versus coordinate-space.

**Comment by delandmeterp** [comment]
Hi all,
Yes as you said, the confusion comes from the fact that for B and C grids, the grid associated to a field does is not located at the position of the field nodes, i.e. U[k, j, i] is not located at x[k, j, i].

A way to clarify it would be to put directly in the documentation figure 2.B of the GMD paper, and say explicitly (with indexing) where the U, V, W nodes are located, like @erikvansebille has written in the new pynb.

**Comment by delandmeterp** [comment]
@erikvansebille, reviewing on github a notebook file is a bit ugly. 
I'll rather send you an annotated pdf

**Comment by delandmeterp** [comment]
What do you think of adding one more section to the doc? Something like this

## How does Parcels deal with B and C grids?

To keep it simple, let's take the case of a 2D C grid.

Why doesn't Parcels need the exact location of the U and V nodes? This is because in C grids, U and V nodes are always located the same way: for each grid cell, U nodes are located at the middle of the left and right edges, and V nodes are located at the bottom and top edges (insert here figure 2B of the paper). The only information that Parcels needs is the location of the 4 corners of the cell, which are called the f-nodes. Those are the one provided by U.grid (which is equal to V.grid).
Now, as you've noticed on the grid illustrated on the figure, there are 4x4 cells. The grid providing the cell corners is a 5x5 grid.
But there are only 4x5 U nodes and 5x4 V nodes, since the grids are staggered.

The indexing notation chosen in Parcels is the same as for Nemo, U[j, i] is located between the cell corners [j-1, i] and [j, i], and  V[j, i] is located between cell corners [j, i-1] and [j, i]This implies that the first row of U data and the first column of V data is never used (and do not physically exist), but the U and V fields are now correctly provided on a 5x5 table. If your original data are provided on a 4x5 U grid and a 5x4 V grid, you need to regrid your table to follow Parcels notation before loading your into it!

For 3D C grids and B grids, the idea is the same. It is important to follow the indexing notation, which is defined in Parcels and in the [GMD paper]. Note that Parcels take care of loading this data for you, through the from_nemo, .... routines.


What do you think? I would say it is a bit technical, but it could give a brief explaination for people who know what they are doing, but are confused by Parcels notations. 
But, if it adds complexity instead fo simplyfying, it's maybe better to ignore it :-)

**Comment by angus-g** [comment]
Thanks @delandmeterp, this helps clarify it for me. Particularly the requirements for regridding to follow the Parcels convention — these kind of fiddly things are nice to have written down!

**Comment by erikvansebille** [comment]
Fixed in #822
